src/clj_irods/icat.clj

```diff
@@ -0,0 +1,207 @@
+    (reduce
+      (fn [m ks]
+        (let [ks4            (take 4 ks)
+              [limit offset] (drop 4 ks)]
+          (if-let [extant (get-in m ks4)]
+            (assoc-in m ks4 (merge-listing extant (get-in listings ks) limit offset))
+            (assoc-in m ks (get-in listings ks)))))
+      {}
+      key-paths)))
+
+(defn get-range
+  "Given merged listings selected down to limit-offset section, get a section of results for a limit & offset"
+  [merged-listings limit offset]
+  (let [all-ranges (mapv (partial take 2) (keys-in merged-listings))
+        unbounded-limit-offset (filter (fn [[l o]] (and (nil? l) (>= offset o))) all-ranges)
+        bounded-limit-offset (and
+                               (not (nil? limit))
+                               (filter (fn [[l o]] (and (not (nil? l)) (<= limit l) (>= offset o))) all-ranges))
+        finished-ranges (filter (fn [[l o]] (and (not (nil? l)) (>= offset o) (> l (count (get-in merged-listings [l o]))))) all-ranges)]
+    (cond
+      ;; we have a cached listing with (:limit nil) and an offset <= our requested offset
+      (seq unbounded-limit-offset)
+      (let [[cached-limit cached-offset] (first unbounded-limit-offset)
+            results (get-in merged-listings [cached-limit cached-offset])]
+        (if (nil? limit)
+          (delay (drop (- offset cached-offset) results))
+          (delay (take limit (drop (- offset cached-offset) results)))))
+
+      ;; we have a cached listing where the number of items is less than its limit (i.e. -- it reached the end) and an offset <= our requested offset
+      (seq finished-ranges)
+      (let [[cached-limit cached-offset] (first finished-ranges)
+            results (get-in merged-listings [cached-limit cached-offset])]
+        (if (nil? limit)
+          (delay (drop (- offset cached-offset) results))
+          (delay (take limit (drop (- offset cached-offset) results)))))
+
+      ;; we have a bounded limit and offset in the request, and a cached entry that can fulfill it
+      (and (not (nil? limit)) (seq bounded-limit-offset))
+      (let [[cached-limit cached-offset] (first bounded-limit-offset)
+            results (get-in merged-listings [cached-limit cached-offset])]
+        (delay (take limit (drop (- offset cached-offset) results))))
+      
+      :else
+      nil)))
+
+;; cached offset 10 limit 10, request offset 15 limit 5 vs limit 10
+;; 10 - (15 - 10) = 5 = maximum limit we can do, i.e. must be <= request limit
+;; cached offset 0 limit 1000; request offset 100 limit 100
+;; (1000 - (100 - 0) = 900)
+
+;; listings are {limit {offset [seq]}}
+;; our offset must be <= the offset of the merged+cached listing, regardless
+;; if our limit is nil, then either the cached limit must be nil, or (cached-limit + cached-offset > actual # of items in cached listing)
+;; if our limit is specified, then (cached-limit - (our-offset - cached-offset)) >= our-limit, or cached-limit must be nil
+
+(defn flatten-cached-listings
```

Comment:
May still replace this with the `comp`-based version that came up in Slack. It's more concise, but I can't decide if it's clearer.